### PR TITLE
Add range and partition support on FoundationDB adapter for batch processing

### PR DESCRIPTION
**Pull request by ruweih**

### CODE REVIEW

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java [hunk 1]
```diff
@@ -235,6 +238,36 @@ public void delete(StaticBuffer key, StoreTransaction txh) throws BackendExcepti
...
     }
+    public List<StaticBuffer> getBoundaryKeys() {
+        List<StaticBuffer> keys = new ArrayList<>();
+        try (CloseableAsyncIterator<byte[]> it = LocalityUtil.getBoundaryKeys(manager.db, db.range().begin, db.range().end)) {
+            it.forEachRemaining(key -> keys.add(getBuffer(db.unpack(key).getBytes(0))));
+        }
+        return keys;
+    }
+
+    public Iterator<KeyValueEntry> getSliceAsync(byte[] keyStart, byte[] keyEnd, int limit, KeySelector selector, StoreTransaction txh) throws BackendException {
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java:249]: For convenience, `getSliceAsync` should have the same signature as `getSlice`.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 2]
```diff
@@ -190,6 +190,10 @@ private TransactionClose(String msg) {
         return result;
     }
 
+    public Iterable<KeyValue> getRangeAsync(final byte[] startKey, final byte[] endKey, final int limit) throws PermanentBackendException {
+        return tx.getRange(startKey, endKey, limit);
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:194]: This method will block until the full list of results is ready, which is not what one would expect from the method name.
I would prefer to move the logic of the `getRange` method here and design it so that it runs in a properly async manner. Then all that's left to do is call `getRangeAsync` within `getRange` and additionally wait for it's completion before returning the results.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java [hunk 3]
```diff
@@ -235,6 +238,36 @@ public void delete(StaticBuffer key, StoreTransaction txh) throws BackendExcepti
             throw new PermanentBackendException(e);
         }
     }
+    public List<StaticBuffer> getBoundaryKeys() {
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java:241]: Where is this method called? Is it somewhere inside JanusGraph where the `OrderedKeyValueStore` is used?

**Review by rngcntr — CHANGES_REQUESTED** [review]
Thanks for your contribution @ruweih!
Using FoundationDB's [async reads](https://apple.github.io/foundationdb/javadoc/com/apple/foundationdb/ReadTransaction.html) is a very nice idea. But actually, I don't really know where `getSliceAsync` will be used in JanusGraph. Is it used at all?

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 2 above)
Comment (ruweih) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:194]: From the method signature , it seems return an async result, that's why I choose the name for this method:
https://apple.github.io/foundationdb/javadoc/com/apple/foundationdb/ReadTransaction.html#getRange-byte:A-byte:A-int-

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java (hunk 1 above)
Comment (ruweih) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java:249]: Maybe, but we can not use the inner class "FoundationDBRecordIterator" since it requires data been retrieved first, which defeat the purpose of this method to not retrieving data in memory first.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java (hunk 3 above)
Comment (ruweih) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java:241]: As I mentioned in PR, those method are not called within JG, it's used for bulk access in Spark jobs.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 2 above)
Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:194]: Huh, I'm pretty confident at the time I wrote this comment, there were both `getRange` and `getRangeAsync` available in FDB. But now my comment is obsolete of course :smile:

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java (hunk 1 above)
Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java:249]: FoundationDB offers an [`AsyncIterable`](https://apple.github.io/foundationdb/javadoc/com/apple/foundationdb/async/AsyncIterable.html) which extends `Iterable`. You could maybe modify the `FoundationDBRecordIterator` to not take a `List` in it's constructor but any `Iterable`. This way it would be possible to use the `FoundationDBRecordIterator` for async results.

**Comment by pluradj** [comment]
@ruweih Regarding the CLA check, please fix your commits so that you sign them with your IBM email address instead of your Gmail address. Thanks.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java (hunk 1 above)
Comment (ruweih) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java]: getSliceAsync is introduced from other PR recently, so no need this method any more.

**Comment by porunov** [comment]
@ruweih Thank you for the contribution! You have 4 commits in this PR and two of them are not signed properly. If you squash and sign you commit, EasyCLA will pass. Let me know if you have troubles squashing those commits and I will help

**Comment by ruweih** [comment]
> @ruweih Regarding the CLA check, please fix your commits so that you sign them with your IBM email address instead of your Gmail address. Thanks.

How could we change the existing commits?

**Comment by ruweih** [comment]
> @ruweih Thank you for the contribution! You have 4 commits in this PR and two of them are not signed properly. If you squash and sign you commit, EasyCLA will pass. Let me know if you have troubles squashing those commits and I will help

Thanks. I used git desktop to merge from master, which for some reason does not signed it properly, how do I change the signature of those existing commits?

**Comment by porunov** [comment]
> Thanks. I used git desktop to merge from master, which for some reason does not signed it properly, how do I change the signature of those existing commits?

@ruweih There are different ways to do it. In you situation, you have some merge commits. You could use `git rebase` to pick one commit and squash other commits in it ( i.e. `git rebase -i 4408551d42c3fafeeee70f103f28fefb409022d3`) but in your situation you will also need to resolve conflicts and then do `git rebase --continue` after the resolutions.

In some cases you know that you PR changes are good and you don't want to worry about past commits and resolve merge commits conflicts. If so, you can merge you branch into the target branch with a squash and then reset you feature branch from target branch, push it and then reset you local target branch from origin (or anything else to remove you squashed merge commit).
So, basically, to simplify current squash process I would basically do something like:
```
git checkout master
git merge --squash rangekeys-ruweih
git commit -m "Add range and partition support on FoundationDB adapter for batch processing" -s
git checkout rangekeys-ruweih
git reset --hard master
git push -f
git checkout master
git reset --hard origin/master
```
That said, you can use any other technique to squash your commits but make sure to sign them properly with `-s` parameter.

**Comment by ruweih** [comment]
@porunov Thanks for the help. All checks been passed after the fix.

@rngcntr The updated PR removed the need for own async iterator, which should address your previous comment. Please review again and clean the change request.

**Review by rngcntr — APPROVED** [review]
LGTM

**Comment by ruweih** [comment]
@rngcntr Seems I don't have the write permission to merge it. Could you or someone else please merge the PR in your convenience if there is no more issue need to be addressed?

**Comment by porunov** [comment]
@ruweih - `two committer approvals` OR `one committer approval + 1 week for lazy consensus` is needed to merge the PR

**Comment by theZiki** [comment]
Can somebody merge this, I am waiting for this fix for one year :-)

**Comment by rngcntr** [comment]
@theZiki We need either another committer to approve the PR or wait another four days until we are good to go.
Although you are not currently a committer, you are welcome to review the PR, making the review process easier for other committers. 
(Just in general, probably not as relevant for this relatively small change)

**Comment by davisdk** [comment]
I've also got my eyes on this one. :] Any chance we can get it merged shortly? Thanks!

**Comment by rngcntr** [comment]
@davisdk Thanks for reminding me. Since there were no remarks by others, we can merge this PR now.
